Prevent lack of provider JSON blocking POST requests

gregcorbett opened the pull request:
The lack of an available Indigo CMDB will cause POST requests to be denied even when the requester has been explicitly granted access in the `ALLOWED_TO_POST` variable.

This is caused by a failure to retrieve the provider JSON from an Indigo CMDB, which causes the method that determines whether the signer is valid (and by extension, whether the POST request should be authorized) to return false regardless of whether the signer has also been granted special access.

As the example Indigo CMDB is no longer accessible, all POST requests get denied.

This commit fixes that by making the method continue on to check the `ALLOWED_TO_POST` variable if the list of providers can't be retrieved.

gregcorbett committed:
Prevent lack of provider JSON blocking all access

- If the provider JSON cannot be retrieved
  from an Indigo CMDB, then the method that
  determines wether the signer is valid (and
  by extension, wether the POST request should
  be authorized) returns false regardless of
  wether the signer has also been
  granted special access.
- As the example Indigo CMDB is no longer
  accessible, all POST requests get denied.
- This commit fixes that by making the method
  continue on to check the ALLOWED_TO_POST
  variable if the list of provders can't be
  retrieved

tofu-rocketry reviewed (commented):
Seems like bit of a fudge. Wouldn't it be better to do the ALLOWED_TO_POST check first, rather than setting this list to empty and taking advantage of the fact that it will fall through? I feel that there should only be a single fall through at the end (as there already is).

gregcorbett commented:
It could do it that way. The reason I left it was to maintain the order in which the credentials are checked - banned, listed provider, special case (strongest assertion to weakest)

tofu-rocketry commented:
Could we get a third opinion on this @jrha, please?

jrha commented:
IMHO it depends whether you view `ALLOWED_TO_POST` as a special case, or an override.

Should `ALLOWED_TO_POST` override banned?

It feels like an override to me, in which case checking it first seems like the right thing to do. i.e. there is no point checking anything else as the poster is always allowed to post.

gregcorbett committed:
Rename _get_provider_list to be more descriptive

- as the method infact returns INDIGO provider information
  from the INDIGO CMDB as JSON.
- for other sources of providers this may be different.
- this commit also renames some method variables to match
  the method name change
- also alter tests to mock the newly renamed method

gregcorbett committed:
Refactor parsing of INDIGO CMDB JSON

- have it return a list of hostnames, even if that
  list is empty because it was not able to extract
  any hostnames

gregcorbett committed:
Refactor INDIGO provider handling into own method

- so if there are any future ways of registering as an
  INDIGO provider, only the new method needs to change

gregcorbett commented:
@tofu-rocketry, I have re-factored this a bit following our discussions this morning (to the point that the original commit code be rebased away). 

It preserves the order of checks, but factors out all retrieving provider json and extracting hostnames into separate methods so that in the future we can more easily support other ways of registering providers.

tofu-rocketry commented on api/views/CloudRecordView.py:
```diff
@@ -139,6 +132,29 @@ def _signer_is_valid(self, signer_dn):
+        # Return any hostnames we can parse from the provider JSON.
+        return self._parse_hostnames_indigo_cmdb(provider_json)
+
+    def _signer_is_valid(self, signer_dn):
```
This function's a lot easier to follow now. :)

tofu-rocketry commented on api/views/CloudRecordView.py:
```diff
@@ -139,6 +132,29 @@ def _signer_is_valid(self, signer_dn):
+        if signer in self._get_indigo_providers():
+            return True
+
         if signer_dn in settings.ALLOWED_TO_POST:
```
I think we should incorporate James' comment now and put the this before the indigo providers check.

gregcorbett replied:
changed

tofu-rocketry commented on api/views/CloudRecordView.py:
```diff
@@ -139,6 +132,29 @@ def _signer_is_valid(self, signer_dn):
+            return False
+
+        if signer in self._get_indigo_providers():
+            return True
```
Should this at least log to DEBUG?

gregcorbett replied:
It probably should. I put it in as `info` to match the other returns in this method

tofu-rocketry reviewed (changes requested):
Bit of an messy diff, but easier to follow overall just looking at the full file.

gregcorbett committed:
Log when access is granted via the provider list

- log level matches the other similar logging
  in this method

gregcorbett committed:
Prevent needless provider checks

- if a hosts DN is listed in ALLOWED_TO_POST, they will be
  allowed to POST regardless of whether they are listed as
  as a provider or not. So why not check ALLOWED_TO_POST
  first and avoid a potentially expensive provider check
- this changes preserves the fact that BANNED_FROM_POST
  overwrites both ALLOWED_TO_POST and any provider lists

gregcorbett commented:
changes made

tofu-rocketry commented:
Could you rebase this so I can do the final approval, please?

gregcorbett commented:
rebased
